Detectors/GlobalTrackingWorkflow/tpcinterpolationworkflow/src/TPCInterpolationSpec.cxx

```diff
@@ -0,0 +1,259 @@
+    LOG(ERROR) << "Reading bufferized TPC clusters, this should not happen";
+    // some of the inputs have been buffered
+    for (size_t sector = 0; sector < sectorStatus.size(); ++sector) {
+      if (sectorStatus.test(sector)) {
+        clustersTPC[sector] = gsl::span(&mBufferedTPCClusters[sector].front(), mBufferedTPCClusters[sector].size());
+      }
+    }
+  }
+
+  // Just print TPC clusters status
+  {
+    // make human readable information from the bitfield
+    std::string bitInfo;
+    auto nActiveBits = validSectors.count();
+    if (((uint64_t)0x1 << nActiveBits) == validSectors.to_ulong() + 1) {
+      // sectors 0 to some upper bound are active
+      bitInfo = "0-" + std::to_string(nActiveBits - 1);
+    } else {
+      int rangeStart = -1;
+      int rangeEnd = -1;
+      for (size_t sector = 0; sector < validSectors.size(); sector++) {
+        if (validSectors.test(sector)) {
+          if (rangeStart < 0) {
+            if (rangeEnd >= 0) {
+              bitInfo += ",";
+            }
+            bitInfo += std::to_string(sector);
+            if (nActiveBits == 1) {
+              break;
+            }
+            rangeStart = sector;
+          }
+          rangeEnd = sector;
+        } else {
+          if (rangeStart >= 0 && rangeEnd > rangeStart) {
+            bitInfo += "-" + std::to_string(rangeEnd);
+          }
+          rangeStart = -1;
+        }
+      }
+      if (rangeStart >= 0 && rangeEnd > rangeStart) {
+        bitInfo += "-" + std::to_string(rangeEnd);
+      }
+    }
+    LOG(INFO) << "running matching for sector(s) " << bitInfo;
+  }
+
+  o2::tpc::ClusterNativeAccess clusterIndex;
+  std::unique_ptr<o2::tpc::ClusterNative[]> clusterBuffer;
+  o2::tpc::MCLabelContainer clusterMCBuffer;
+  memset(&clusterIndex, 0, sizeof(clusterIndex));
+  o2::tpc::ClusterNativeHelper::Reader::fillIndex(clusterIndex, clusterBuffer, clusterMCBuffer, clustersTPC, mcInputs, [&validSectors](auto& index) { return validSectors.test(index); });
+  //----------------------------<< TPC Clusters loading <<------------------------------------------
+
+  // pass input data to TrackInterpolation object
+  mInterpolation.setITSTracksInp(tracksITS);
+  mInterpolation.setTPCTracksInp(tracksTPC);
+  mInterpolation.setTPCTrackClusIdxInp(tracksTPCClRefs);
+  mInterpolation.setTPCClustersInp(&clusterIndex);
+  mInterpolation.setTOFMatchesInp(trackMatchesTOF);
+  mInterpolation.setITSTPCTrackMatchesInp(tracksITSTPC);
+  mInterpolation.setTOFClustersInp(clustersTOF);
+
+  printf("TPC Interpolation Workflow initialized. Start processing...\n");
+
+  mInterpolation.process();
+
+  pc.outputs().snapshot(Output{"GLO", "TPCINT_TRK", 0, Lifetime::Timeframe}, mInterpolation.getReferenceTracks());
+  pc.outputs().snapshot(Output{"GLO", "TPCINT_RES", 0, Lifetime::Timeframe}, mInterpolation.getClusterResiduals());
+  if (mUseMC) {
+    // TODO
+  }
+  mFinished = true;
+  pc.services().get<ControlService>().endOfStream();
```

Comment:
you don't need endOfStream and readyToQuit here, the completion will be steered by devices which feed the data.

Reply:
Should I also remove mFinished here or is it needed?

Reply:
Yes, the processor device should not decide if the processing is finished or not, this will be done upstream.